Gossiper: replace seastar threads with coroutines

bhalevy opened the pull request:
Many of the gossiper internal functions currently use seastar threads for historical reasons,
but since they are short living, the cost of spawning a seastar thread for them is excessive
and they can be simplified and made more efficient using coroutines.

bhalevy commented:
I'd like to merge this PR before making further changes in those areas for https://github.com/scylladb/scylladb/issues/12282

avikivity commented on gms/gossiper.cc:
```diff
@@ -2172,32 +2173,31 @@ future<> gossiper::add_local_application_state(std::initializer_list<std::pair<a
-        return seastar::async([g = gossiper.shared_from_this(), states = std::move(states)]() mutable {
-            auto& gossiper = *g;
+    try {
+        co_await container().invoke_on(0, [&] (gossiper& gossiper) mutable -> future<> {
```
coroutine::lambda()?

bhalevy replied:
I understand we need coroutine::lambda for continuations.
Is it needed for invoke_on / submit_to calls as well?

Is it needed when the lambda has no captures?

avikivity replied:
It's capturing &, so one variable.

I looked at submit_to() and it moves the lambda to a safe place, so it's not in the same category as a continuation. We should mark it as safe for continuations.

avikivity commented on gms/gossiper.cc:
```diff
@@ -214,25 +217,27 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
                     }
-                }).handle_exception([this, from] (std::exception_ptr ep) {
+        } catch (...) {
+            ep = std::current_exception();
```
Can't you declare ep here?

avikivity commented on gms/gossiper.cc:
```diff
@@ -214,25 +217,27 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
                         p.syn_msg = {};
                     }
                     logger.warn("Failed to process gossip syn msg from node {}:  {}", from, ep);
-                    return make_exception_future<stop_iteration>(ep);
```
And return it here?

bhalevy replied:
I'll look into it.

avikivity commented on gms/gossiper.cc:
```diff
@@ -337,27 +341,28 @@ future<> gossiper::handle_ack_msg(msg_addr id, gossip_digest_ack ack_msg) {
                             p.ack_msg_digest = {};
                             logger.warn("Failed to process gossip ack msg digests from node {}: {}", from, ep);
                         }
-                        return make_exception_future<stop_iteration>(ep);
```
Similarly, here. Can't ep live in the catch clause? Makes everything simpler to read later.

avikivity commented on gms/gossiper.cc:
```diff
@@ -1305,10 +1305,10 @@ future<> gossiper::unsafe_assassinate_endpoint(sstring address) {
 future<> gossiper::assassinate_endpoint(sstring address) {
-    return container().invoke_on(0, [address] (auto&& gossiper) {
-        return seastar::async([&gossiper, g = gossiper.shared_from_this(), address] {
+    co_await container().invoke_on(0, [&] (auto&& gossiper) -> future<> {
```
coroutine::lambda?

avikivity commented:
Looks good, some minor comments.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3600/

bhalevy commented:
v2 ([47dc287](https://github.com/scylladb/scylladb/commit/47dc287efd4f33a63b91fd9d77959f846bf3c3ed)):
- cleaned up includes that were added by the editor
- define `auto ep = std::current_exception()` in catch clauses, then throw
- added: `gossiper: handle_ack_msg: always log warning on exception`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3622/
